pkg/sql/schemachanger/eav/types.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package eav
+
+import "strconv"
+
+// Type is an enumerations of the types exported by this package.
+type Type int
+
+//go:generate stringer --type Type --trimprefix Type
+
+// Enumeration of the types exported by this package.
+const (
+	_ Type = iota
+	TypeInt64
+	TypeInt32
+	TypeUint32
+	TypeString
+
+	NumTypes int = iota
+)
+
+// Types exported by this package.
+type (
+	// Int64 is a value representing an int64.
+	Int64 int64
+	// Int32 is a value representing an int32.
+	Int32 int32
+	// Uint32 is a value representing a uint32.
+	Uint32 uint32
+	// String is a value representing a string.
+	String string
+)
```

Comment:
Instead of this approach have you considered another in which all values are interned? To me it seems like it would have several advantages:
- for schema changes at least the universe of values doesn't seem like it would be large
- eq comparisons and type checks/lookups become easy-peasy and this eav code is going to be doing a lot of these
- it would also play well with sync pools

I played a bit with the following and it worked well: 
```go
type Value int

func (v Value) String() string {
	return internedString(v.Type(), v.idx())
}

func (v Value) Type() Type {
	return Type(int(v) & 15)
}

func (v Value) IsNil() bool {
	return int(v)&16 == 0
}

func (v Value) idx() int {
	return int(v) >> 5
}

func (v Value) Compare(other Value) (ok, less, eq bool) {
	if v == other {
		return true, false, true
	}
	t1 := v.Type()
	t2 := other.Type()
	if t1 != t2 {
		return false, false, false
	}
        // TODO deal with IsNil()
	return true, internedLess(t1, v.idx(), other.idx()), false
}

func NilValue(t Type) Value {
	return Value(t)
}

var interned struct {
	int64map    map[int64]Value
	int64values []int64
...
}

func InternInt64(uninterned int64) Value {
	v, ok := interned.int64map[uninterned]
	if ok {
		return v
	}
	idx := len(interned.int64values)
	interned.int64values = append(interned.int64values, uninterned)
	v = Value((idx << 5) | 16 | int(TypeInt64))
	interned.int64map[uninterned] = v
	return v
}

func InternedInt64(v Value) int64 {
	if v.IsNil() {
		panic("nil")
	}
	if v.Type() != TypeInt64 {
		panic("type")
	}
	return interned.int64values[v.idx()]
}

func internedString(t Type, idx int) string {
	switch t {
	case TypeInt64:
		return strconv.FormatInt(interned.int64values[idx], 10)
...
	}
}

func internedLess(t Type, idx1, idx2 int) bool {
	switch t {
	case TypeInt64:
		return interned.int64values[idx1] < interned.int64values[idx2]
...
	}
}
```